src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationIconValueProvider.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.IO;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties.Package
+{
+    // Sets the ApplicationIcon property to the path to the .ico file. This path is relative
+    // to the project file directory. If the selected file is not within the project directory
+    // tree, the file will be copied into the project file directory. Previously copied files
+    // are not deleted when a new file is selected. ApplicationIcon allows for both full paths
+    // and relative paths anywhere on disk, but we are not utilizing that in this implementation.
+    // The icon file does not need to be included in the project (such as a None/Content item).
+    [ExportInterceptingPropertyValueProvider(ConfigurationGeneral.ApplicationIconProperty, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class ApplicationIconValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IFileSystem _fileSystem;
+
+        [ImportingConstructor]
+        public ApplicationIconValueProvider(UnconfiguredProject unconfiguredProject, IFileSystem fileSystem)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _fileSystem = fileSystem;
+        }
+
+        public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            // Allows the user to clear the application icon.
+            if (string.IsNullOrEmpty(unevaluatedPropertyValue))
+            {
+                return null;
+            }
+
+            // If selected path is invalid, ignore and return previous value.
+            if (!File.Exists(unevaluatedPropertyValue))
```

Comment:
This is the unevaluated property value, this could be "$(FooBar)" you can't check to see if it exists.

Reply:
This code only runs when setting this property thought the new property pages UI. It is reasonable to assume that people using the UI are not using MSBuild variables. The UI drives you toward using the `Browse` button to select a file. When used, it will insert the full filepath on disk to the page, and that will be passed into this method. So, the majority use-case is a full-path string.

Currently, there isn't an API to have a value provider set the value *after* evaluation has occurred. I'm assuming that could be added in the future. There's possibly a way to use the evaluated value Get method to do this but that's not the normal way to set property values in these new pages.